File: implementations/go/internal/infrastructure/postgres/order_repository.go
package postgres

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/google/uuid"
	"github.com/jmoiron/sqlx"

	"github.com/example/orders-service/internal/domain"
)

type OrderRepository struct {
	db *sqlx.DB
}

func NewOrderRepository(db *sqlx.DB) *OrderRepository {
	return &OrderRepository{db: db}
}

func (r *OrderRepository) Create(ctx context.Context, order *domain.Order) error {
	tx, err := r.db.BeginTxx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback()

	query := `INSERT INTO orders (id, customer_id, status, total_amount, created_at, updated_at)
		VALUES ($1, $2, $3, $4, $5, $6)`
	_, err = tx.ExecContext(ctx, query,
		order.ID, order.CustomerID, order.Status,
		order.TotalAmount, order.CreatedAt, order.UpdatedAt,
	)
	if err != nil {
		return fmt.Errorf("insert order: %w", err)
	}

	itemQuery := `INSERT INTO order_items (id, order_id, product_id, quantity, unit_price)
		VALUES ($1, $2, $3, $4, $5)`
	for _, item := range order.Items {
		_, err = tx.ExecContext(ctx, itemQuery,
			item.ID, item.OrderID, item.ProductID,
			item.Quantity, item.UnitPrice,
		)
		if err != nil {
			return fmt.Errorf("insert order item: %w", err)
		}
	}

	if err := tx.Commit(); err != nil {
		return fmt.Errorf("commit transaction: %w", err)
	}

	return nil
}

func (r *OrderRepository) GetByID(ctx context.Context, id uuid.UUID) (*domain.Order, error) {
	var order domain.Order
	query := `SELECT id, customer_id, status, total_amount, created_at, updated_at
		FROM orders WHERE id = $1`
	err := r.db.GetContext(ctx, &order, query, id)
	if err != nil {
		if err == sql.ErrNoRows {
			return nil, domain.ErrOrderNotFound
		}
		return nil, fmt.Errorf("get order: %w", err)
	}

	items, err := r.getItemsByOrderID(ctx, id)
	if err != nil {
		return nil, err
	}
	order.Items = items

	return &order, nil
}

func (r *OrderRepository) List(ctx context.Context, filter domain.OrderFilter) ([]*domain.Order, int, error) {
	where := "WHERE 1=1"
	args := []interface{}{}
	argIdx := 1

	if filter.CustomerID != nil {
		where += fmt.Sprintf(" AND customer_id = $%d", argIdx)
		args = append(args, *filter.CustomerID)
		argIdx++
	}
	if filter.Status != nil {
		where += fmt.Sprintf(" AND status = $%d", argIdx)
		args = append(args, *filter.Status)
		argIdx++
	}

	countQuery := fmt.Sprintf("SELECT COUNT(*) FROM orders %s", where)
	var total int
	if err := r.db.GetContext(ctx, &total, countQuery, args...); err != nil {
		return nil, 0, fmt.Errorf("count orders: %w", err)
	}

	offset := (filter.Page - 1) * filter.Limit
	query := fmt.Sprintf(`SELECT id, customer_id, status, total_amount, created_at, updated_at
		FROM orders %s ORDER BY created_at DESC LIMIT $%d OFFSET $%d`,
		where, argIdx, argIdx+1)
	args = append(args, filter.Limit, offset)

	var orders []*domain.Order
	if err := r.db.SelectContext(ctx, &orders, query, args...); err != nil {
		return nil, 0, fmt.Errorf("list orders: %w", err)
	}

	for _, order := range orders {
		items, err := r.getItemsByOrderID(ctx, order.ID)
		if err != nil {
			return nil, 0, err
		}
		order.Items = items
	}

	return orders, total, nil
}

func (r *OrderRepository) UpdateStatus(ctx context.Context, id uuid.UUID, status domain.OrderStatus) error {
	query := `UPDATE orders SET status = $1, updated_at = NOW() WHERE id = $2`
	result, err := r.db.ExecContext(ctx, query, status, id)
	if err != nil {
		return fmt.Errorf("update order status: %w", err)
	}
	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("get rows affected: %w", err)
	}
	if rows == 0 {
		return domain.ErrOrderNotFound
	}
	return nil
}

func (r *OrderRepository) Delete(ctx context.Context, id uuid.UUID) error {
	tx, err := r.db.BeginTxx(ctx, nil)
	if err != nil {
		return fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback()

	_, err = tx.ExecContext(ctx, `DELETE FROM order_items WHERE order_id = $1`, id)
	if err != nil {
		return fmt.Errorf("delete order items: %w", err)
	}

	result, err := tx.ExecContext(ctx, `DELETE FROM orders WHERE id = $1`, id)
	if err != nil {
		return fmt.Errorf("delete order: %w", err)
	}
	rows, err := result.RowsAffected()
	if err != nil {
		return fmt.Errorf("get rows affected: %w", err)
	}
	if rows == 0 {
		return domain.ErrOrderNotFound
	}

	return tx.Commit()
}

func (r *OrderRepository) getItemsByOrderID(ctx context.Context, orderID uuid.UUID) ([]domain.OrderItem, error) {
	var items []domain.OrderItem
	query := `SELECT id, order_id, product_id, quantity, unit_price
		FROM order_items WHERE order_id = $1`
	if err := r.db.SelectContext(ctx, &items, query, orderID); err != nil {
		return nil, fmt.Errorf("get order items: %w", err)
	}
	return items, nil
}

File: implementations/scala/src/main/scala/orders/infrastructure/OrderRepository.scala
package orders.infrastructure

import cats.effect.IO
import cats.syntax.all.*
import doobie.*
import doobie.implicits.*
import doobie.postgres.implicits.*
import orders.domain.{Order, OrderItem}

import java.time.OffsetDateTime
import java.util.UUID

final class OrderRepository(xa: Transactor[IO]):
  def find(id: UUID): IO[Option[Order]] =
    val order =
      sql"select id, customer_id, status, total_amount::float8, created_at, updated_at from orders where id = $id"
        .query[(UUID, UUID, String, Double, OffsetDateTime, OffsetDateTime)]
        .option
    val items =
      sql"select id, order_id, product_id, quantity, unit_price::float8 from order_items where order_id = $id order by created_at, id"
        .query[OrderItem]
        .to[List]

    (order, items)
      .mapN((row, orderItems) => row.map(o => Order(o._1, o._2, o._3, o._4, o._5, o._6, orderItems)))
      .transact(xa)

  def updateStatus(id: UUID, status: String): IO[Int] =
    sql"update orders set status = $status, updated_at = now() where id = $id".update.run.transact(xa)

  def list(customerId: Option[UUID], status: Option[String], page: Int, limit: Int): IO[(Int, List[Order])] =
    val base = fr"from orders where 1=1" ++ customerId.fold(fr"")(id => fr"and customer_id = $id") ++ status.fold(fr"")(
      value => fr"and status = $value"
    )
    val total = (fr"select count(*)" ++ base).query[Int].unique
    val ids = (fr"select id" ++ base ++ fr"order by created_at desc limit $limit offset ${(page - 1) * limit}").query[
      UUID
    ].to[List]
    (total, ids).mapN((count, orderIds) => (count, orderIds)).transact(xa).flatMap { case (count, orderIds) =>
      orderIds.traverse(find).map(orders => (count, orders.flatten))
    }

  def create(id: UUID, customerId: UUID, items: List[orders.domain.CreateItem]): IO[Unit] =
    val total = items.foldLeft(java.math.BigDecimal.ZERO)((sum, item) =>
      sum.add(item.unitPrice.multiply(java.math.BigDecimal.valueOf(item.quantity.toLong)))
    )
    val order =
      sql"insert into orders (id, customer_id, status, total_amount) values ($id, $customerId, 'pending', $total)"
        .update.run
    val orderItems = items.traverse { item =>
      val itemId = UUID.randomUUID()
      sql"insert into order_items (id, order_id, product_id, quantity, unit_price) values ($itemId, $id, ${item.productId}, ${item.quantity}, ${item.unitPrice})"
        .update.run
    }
    (order *> orderItems).transact(xa).void
